File: ruby/spec/zser/parser_spec.rb
# frozen_string_literal: true

RSpec.describe Zser::Parser do
  context "message.tjson examples" do
    # Integration test against the decoder
    subject(:parser) { described_class.new(Zser::Decoder.new) }

    MessageExample.load_file.each do |example|
      it example.name do
        if example.success
          expect(parser.parse(example.encoded)).to eq true
          result = parser.finish
          expect(result).to be_a Zser::Object
          expect(result).to eql Zser::Object.from_tjson(example.decoded)
        else
          expect { parser.parse(example.encoded) }.to raise_error Zser::ParseError
        end
      end
    end
  end
end

File: js/test/parser.spec.ts
import { suite, test } from "mocha-typescript";
import { expect } from "chai";
import { MessageExample } from "./support/test_vectors";
import { Decoder } from "../src/decoder";
import { Parser } from "../src/parser";

@suite class ParserSpec {
  static examples: MessageExample[];

  static async before() {
    this.examples = await MessageExample.loadAll();
  }

  @test "vectors/messages.tjson (zser message test vectors)"() {
    for (let example of ParserSpec.examples) {
      let parser = new Parser(new Decoder);

      if (example.success) {
        parser.parse(example.encoded);
        expect(parser.finish()).to.eql(example.decoded);
      } else {
        expect(() => parser.parse(example.encoded)).to.throw;
      }
    }
  }
}
